File: mobile_client/lib/widgets/status_pill.dart
import 'package:flutter/material.dart';

import '../theme.dart';

/// A compact status chip: a coloured label with a soft tinted background, used
/// across the More-tab screens for consistent status presentation. When [live]
/// is true it shows a gently pulsing leading dot (e.g. a running task).
class StatusPill extends StatelessWidget {
  const StatusPill(
    this.label, {
    super.key,
    required this.color,
    this.live = false,
    this.dense = false,
  });

  final String label;
  final Color color;
  final bool live;
  final bool dense;

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(
          horizontal: dense ? 7 : 9, vertical: dense ? 2 : 3),
      decoration: BoxDecoration(
        color: color.withValues(alpha: 0.15),
        borderRadius: BorderRadius.circular(999),
        border: Border.all(color: color.withValues(alpha: 0.35), width: 0.8),
      ),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          if (live) ...[
            PulsingDot(color: color, size: dense ? 6 : 7),
            const SizedBox(width: 6),
          ],
          Text(
            label,
            style: TextStyle(
              color: color,
              fontSize: dense ? 10 : 11,
              fontWeight: FontWeight.w700,
              letterSpacing: 0.4,
            ),
          ),
        ],
      ),
    );
  }
}

/// A small dot that breathes (opacity + halo) — signals a live/active state.
class PulsingDot extends StatefulWidget {
  const PulsingDot({super.key, required this.color, this.size = 8});
  final Color color;
  final double size;

  @override
  State<PulsingDot> createState() => _PulsingDotState();
}

class _PulsingDotState extends State<PulsingDot>
    with SingleTickerProviderStateMixin {
  late final AnimationController _c = AnimationController(
    vsync: this,
    duration: const Duration(milliseconds: 1100),
  )..repeat(reverse: true);

  @override
  void dispose() {
    _c.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _c,
      builder: (_, __) {
        final t = _c.value;
        return Container(
          width: widget.size,
          height: widget.size,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: widget.color,
            boxShadow: [
              BoxShadow(
                color: widget.color.withValues(alpha: 0.25 + 0.45 * (1 - t)),
                blurRadius: 2 + 5 * t,
                spreadRadius: 0.5 + 2.5 * t,
              ),
            ],
          ),
        );
      },
    );
  }
}

/// A bold section heading used inside detail sheets and dashboards.
class SectionHeader extends StatelessWidget {
  const SectionHeader(this.text, {super.key, this.trailing});
  final String text;
  final Widget? trailing;

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 10, top: 4),
      child: Row(
        children: [
          Text(
            text.toUpperCase(),
            style: const TextStyle(
              color: JcTheme.muted,
              fontSize: 11.5,
              fontWeight: FontWeight.w700,
              letterSpacing: 0.8,
            ),
          ),
          const Spacer(),
          if (trailing != null) trailing!,
        ],
      ),
    );
  }
}

File: ios_app/JarvisCopilot/Copilot/UI/StatusPill.swift
import SwiftUI

/// A compact status chip: a coloured label on a soft tinted background, used
/// across the More-tab screens so status reads the same everywhere. `live` adds a
/// gently pulsing leading dot (a running task). Ported from `status_pill.dart`.
struct StatusPill: View {
    let label: String
    let color: Color
    var live: Bool = false
    var dense: Bool = false

    init(_ label: String, color: Color, live: Bool = false, dense: Bool = false) {
        self.label = label
        self.color = color
        self.live = live
        self.dense = dense
    }

    var body: some View {
        // The Voice page's status pill: a quiet translucent capsule where only
        // the dot carries the colour, instead of a tinted badge.
        HStack(spacing: 6) {
            if live { PulsingDot(color: color, size: dense ? 6 : 7) }
            else { Circle().fill(color).frame(width: dense ? 5 : 6, height: dense ? 5 : 6) }
            Text(label)
                .font(.system(size: dense ? 10.5 : 11.5, weight: .medium))
                .foregroundStyle(JcTheme.text.opacity(0.85))
        }
        .padding(.horizontal, dense ? 8 : 11)
        .padding(.vertical, dense ? 3 : 5)
        .background(.white.opacity(0.045), in: Capsule())
    }
}

/// A small dot that breathes (opacity + halo) — signals a live/active state.
///
/// The animation is driven by a single repeating `withAnimation`, so a list of
/// these costs one animation each rather than a ticker per frame.
struct PulsingDot: View {
    let color: Color
    var size: CGFloat = 8

    @State private var expanded = false

    var body: some View {
        Circle()
            .fill(color)
            .frame(width: size, height: size)
            .shadow(color: color.opacity(expanded ? 0.25 : 0.7), radius: expanded ? 3.5 : 1)
            .opacity(expanded ? 0.65 : 1)
            .animation(.easeInOut(duration: 1.1).repeatForever(autoreverses: true), value: expanded)
            .onAppear { expanded = true }
    }
}

/// A bold section heading used inside detail sheets and dashboards.
struct SectionHeader<Trailing: View>: View {
    let text: String
    @ViewBuilder var trailing: Trailing

    init(_ text: String, @ViewBuilder trailing: () -> Trailing) {
        self.text = text
        self.trailing = trailing()
    }

    var body: some View {
        HStack {
            Text(text.uppercased())
                .font(.system(size: 11.5, weight: .bold))
                .kerning(0.8)
                .foregroundStyle(JcTheme.muted)
            Spacer()
            trailing
        }
        .padding(.top, 4)
        .padding(.bottom, 10)
    }
}

extension SectionHeader where Trailing == EmptyView {
    init(_ text: String) { self.init(text) { EmptyView() } }
}
